src/common/hist_util.cc

```diff
@@ -412,25 +451,45 @@ void GHistIndexMatrix::Init(DMatrix* p_fmat, int max_num_bins) {
       }
     }
 
-    index.resize(row_ptr[rbegin + batch.Size()]);
-
-    CHECK_GT(cut.Values().size(), 0U);
+//    index.resize(row_ptr[rbegin + batch.Size()]);
+    const size_t n_disps = cut.Ptrs().size() - 1;
+    if ((max_num_bins - 1 <= static_cast<int>(std::numeric_limits<uint8_t>::max())) && isDense) {
+      index.setBinBound(POWER_OF_TWO_8);
+      index.resize((sizeof(uint8_t)) * row_ptr[rbegin + batch.Size()], n_disps);
+    } else if ((max_num_bins - 1 > static_cast<int>(std::numeric_limits<uint8_t>::max())  &&
+            max_num_bins - 1<= static_cast<int>(std::numeric_limits<uint16_t>::max())) && isDense) {
+      index.setBinBound(POWER_OF_TWO_16);
+      index.resize((sizeof(uint16_t)) * row_ptr[rbegin + batch.Size()], n_disps);
+    } else {
```

Comment:
If ```isDense``` is false, does it mean that it uses ```uint32_t``` always?
I dont understand why

Reply:
uint32_t is always used for build histogram in case `isDense` is false. But for Partition fix  was applied, thanks